rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+
+Preboot plugins have only `setup` and `stop` methods, and can only depend on other preboot plugins. Primary plugins cannot depend on the preboot plugins since Kibana will stop them before starting the primary plugins:
+
+```ts
+export interface PrebootPlugin<TSetup = void, TPluginsSetup extends object = object> {
+  setup(core: CorePrebootSetup, plugins: TPluginsSetup): TSetup;
+  stop?(): void;
+}
+```
+
+To differentiate preboot and primary plugins we'll introduce a new _optional_ `type` property in the plugin manifest. The property can have only two possible values: `preboot` for `preboot` plugins and `primary` for the primary ones. If `type` is omitted, the `primary` value will be assumed.
+
+```json5
+// NOTE(azasypkin): all other existing properties have been omitted for brevity.
+{
+  "type": "preboot", // 'preboot' | 'primary' | undefined
+}
+```
+
+The Plugins service will split plugins into two separate groups during discovery to use them separately at the `preboot`, `setup`, and `start` stages. The Core contract that preboot plugins will receive during their `setup` will be different from the one primary plugins receive, and will only include the functionality that is currently required for the interactive setup mode. We'll discuss this functionality in details in the following sections:
+
+```ts
+export interface CorePrebootSetup {
+  elasticsearch: ElasticsearchServicePrebootSetup;
+  http: HttpServicePrebootSetup;
+  preboot: PrebootServiceSetup;
+}
+```
+
+### 3.2.2 HTTP service
+
+We'll change HTTP service to initialize and start preboot HTTP server (formerly known as `Not Ready` server) in the new `preboot` method instead of `setup`. The returned `InternalHttpServicePrebootSetup` contract will presumably be very similar to the existing `InternalHttpServiceSetup` contract, but will only include APIs we currently need to support interactive setup mode:
+
+```ts
+// NOTE(azasypkin): some existing properties have been omitted for brevity.
+export interface InternalHttpServicePrebootSetup
+  extends Pick<HttpServiceSetup, 'auth' | 'csp' | 'basePath' | 'getServerInfo'> {
+  server: HttpServerSetup['server'];
+  externalUrl: ExternalUrlConfig;
+  registerRoutes(path: string, callback: (router: IRouter) => void): void;
+}
+```
+
+The only part of this contract that will be available to the preboot plugins via `CorePrebootSetup` is the API to register HTTP routes on the already running preboot HTTP server:
+
+```ts
+export interface HttpServicePrebootSetup {
+  registerRoutes(path: string, callback: (router: IRouter) => void): void;
+}
+```
+
+The Core HTTP context available to handlers of the routes registered on the preboot HTTP server will only expose the `uiSettings` service. As explained in the [UI Settings service section](#324-ui-settings-service), this service will give access to the **default** UI settings and their overrides set through Kibana configuration, if any.
```

Comment:
Kibana Plugins register their `ui settings` during the `setup` phase. Thus, we won't be able to have access to plugin ui_settings **defaults** in the "preboot" phase. Thus, users can specify overrides for Core and read Core **defaults** only.

Reply:
Correct, I'll make it clear that we're talking about  `default Core UI settings`, thanks.